### PR TITLE
drivers: Add support for pinctrl in nRF SPI and I2C drivers

### PR DESCRIPTION
**Pull request by anangl**
Add support for the new pinctrl API to the SPI and I2C drivers that handle the following nRF peripherals: SPI, SPIM, SPIS, TWI, and TWIM.
Add also build assertions in the nRF UART drivers that will ensure that every peripheral for which a driver instance is created has some pins assigned to it.

### CODE REVIEW

File: drivers/serial/uart_nrfx_uart.c [hunk 1]
```diff
@@ -1192,6 +1192,13 @@ static int uart_nrfx_pm_action(const struct device *dev,
 PINCTRL_DT_INST_DEFINE(0);
 #endif /* CONFIG_PINCTRL */
 
+BUILD_ASSERT((IS_ENABLED(CONFIG_PINCTRL) &&
+	      DT_PINCTRL_HAS_IDX(DT_DRV_INST(0), 0)) ||
```
Comment (gmarull) [inline drivers/serial/uart_nrfx_uart.c:1196]: entry "sleep" is also required if `CONFIG_PM_DEVICE=y`.

File: samples/boards/nrf/dynamic_pinctrl/boards/nrf52840dk_nrf52840.overlay [hunk 2]
```diff
@@ -17,6 +17,11 @@
 	pinctrl-names = "default", "sleep";
 };
 
+/* Disable uart1 so that it is not left without any pins assigned. */
+&uart1 {
+	status = "disabled";
+};
```
Comment (gmarull) [inline samples/boards/nrf/dynamic_pinctrl/boards/nrf52840dk_nrf52840.overlay:23]: Please add a TODO here about removing this once all boards are ported

File: drivers/spi/spi_nrfx_spi.c [hunk 3]
```diff
@@ -277,8 +291,13 @@ static int spi_nrfx_pm_action(const struct device *dev,
...
 	case PM_DEVICE_ACTION_RESUME:
-		/* No action needed at this point, nrfx_spi_init() will be
-		 * called at configuration before the next transfer.
+		ret = spi_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* nrfx_spi_init() will be called at configuration before
+		 * the next transfer.
```
Comment (gmarull) [inline drivers/spi/spi_nrfx_spi.c:300]: why do who have a helper here? I'd be consistent and use #ifdefs everywhere, as in the sleep case.

File: drivers/spi/spi_nrfx_spi.c [hunk 4]
```diff
@@ -317,9 +344,24 @@ static int spi_nrfx_pm_action(const struct device *dev,
...
+		 .miso_pull = SPI_NRFX_MISO_PULL(idx),))
+
 #define SPI_NRFX_SPI_DEVICE(idx)					       \
-	BUILD_ASSERT(							       \
-		!SPI_PROP(idx, miso_pull_up) || !SPI_PROP(idx, miso_pull_down),\
+	BUILD_ASSERT((IS_ENABLED(CONFIG_PINCTRL) &&			       \
+		      DT_PINCTRL_HAS_IDX(SPI(idx), 0)) ||		       \
+		     (!IS_ENABLED(CONFIG_PINCTRL) &&			       \
+		      DT_NODE_HAS_PROP(SPI(idx), sck_pin)),		       \
+		"No pins assigned to SPI"#idx);				       \
```
Comment (gmarull) [inline drivers/spi/spi_nrfx_spi.c:361]: ditto, sleep case should also be checked. I think creating an intermediate helper macro would improve readability here

File: drivers/spi/spi_nrfx_spim.c [hunk 5]
```diff
@@ -435,8 +450,13 @@ static int spim_nrfx_pm_action(const struct device *dev,
...
 	case PM_DEVICE_ACTION_RESUME:
-		/* No action needed at this point, nrfx_spim_init() will be
-		 * called at configuration before the next transfer.
+		ret = spi_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* nrfx_spim_init() will be called at configuration before
+		 * the next transfer.
```
Comment (gmarull) [inline drivers/spi/spi_nrfx_spim.c:459]: ditto

File: drivers/i2c/i2c_nrfx_twi.c [hunk 6]
```diff
@@ -228,6 +243,11 @@ static int twi_nrfx_pm_action(const struct device *dev,
 
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
+		ret = twim_nrfx_pins_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
```
Comment (gmarull) [inline drivers/i2c/i2c_nrfx_twi.c:249]: ditto

File: drivers/spi/spi_nrfx_spi.c (hunk 3 above)
Comment (anangl) [reply drivers/spi/spi_nrfx_spi.c:300]: Default configuration of the pins needs to be also applied when the driver instance is initialized (in `spi_##idx##_init()`). That's why `spi_nrfx_pins_init()` was introduced.

File: drivers/spi/spi_nrfx_spi.c (hunk 4 above)
Comment (anangl) [reply drivers/spi/spi_nrfx_spi.c:361]: I'll add a generic macro for checking this, to not copy almost the same code in several drivers.

File: drivers/spi/spi_nrfx_spi.c (hunk 3 above)
Comment (gmarull) [reply drivers/spi/spi_nrfx_spi.c:300]: why can't we use regular ifdeffed pinctrl calls? `spi_nrfx_pins_init` naming is confusing IMO, and inconsistent with the sleep counterpart.

Comment (anangl) [reply drivers/spi/spi_nrfx_spi.c:300]: I was afraid that `spi_##idx##_init()`, which is in a macro, would be difficult to read then. But I agree, it is confusing it the current form. I'll get rid of `spi_nrfx_pins_init()`.

File: drivers/spi/spi_nrfx_spi.c (hunk 4 above)
Comment (anangl) [reply drivers/spi/spi_nrfx_spi.c]: Added.

**Review by gmarull — DISMISSED** [review]

**Comment by anangl** [comment]
Rebased.

**Comment by anangl** [comment]
The compliance check fails because of this macro call:
```
		      (FOR_EACH_FIXED_ARG(NRF_DT_ENSURE_NODE_HAS_PROP,	\
					  (||), node_id, __VA_ARGS__))),\
```
but I see no reasonable way of fixing this. The `||` operator must be in parentheses because the macro requires this. For the above form, checkpatch signals an error because of "spaces required around that '||'". For `( || )` in turn, it signals errors because of prohibited spaces after the opening parenthesis and before the closing one...
This could be fixed with:
```
		      (FOR_EACH_FIXED_ARG(NRF_DT_ENSURE_NODE_HAS_PROP, (\
					  ||), node_id, __VA_ARGS__))), \
```
but then the code looks weird and can be confusing, so I'm not sure if that's worth it.

**Comment by gmarull** [comment]
@anangl maybe we can patch checkpatch with something like:

```diff
diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index fb2efb1d09..cc702c91d0 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -4726,6 +4726,13 @@ sub process {
                                                $ok = 1;
                                        }
 
+                                       # some macros require a separator argument to be in parenthesis,
+                                       # e.g. (||).
+                                       if (($op eq '||') &&
+                                           ($ca =~ /\($/ || $cc =~ /^\)/)) {
+                                               $ok = 1;
+                                       }
+
                                        # messages are ERROR, but ?: are CHK
                                        if ($ok == 0) {
                                                my $msg_level = \&ERROR;
```

**Comment by anangl** [comment]
> @anangl maybe we can patch checkpatch with something like:

@gmarull If it is allowed to modify checkpatch this way, that works for me.
Would you mind creating a commit with this patch so that I could include it in the PR, or should I create one with you as the author (as this piece of code is entirely yours)?

**Review by gmarull — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**